packages/@glimmer/runtime/lib/references/dynamic-element.ts

```diff
@@ -0,0 +1,70 @@
+import { InternalComponentCapabilities, InternalComponentManager } from '@glimmer/interfaces';
+import { CurriedType, Dict, Maybe } from '@glimmer/interfaces';
+import { createComputeRef, NULL_REFERENCE, Reference, valueForRef } from '@glimmer/reference';
+import { setInternalComponentManager } from '@glimmer/manager';
+import { curry } from '../curried-value';
+
+const CAPABILITIES: InternalComponentCapabilities = {
+  dynamicLayout: false,
+  dynamicTag: false,
+  prepareArgs: false,
+  createArgs: false,
+  attributeHook: false,
+  elementHook: false,
+  createCaller: false,
+  dynamicScope: false,
+  updateHook: false,
+  createInstance: false,
+  wrapped: false,
+  willDestroy: false,
+  hasSubOwner: false,
+};
+
+export class DynamicElementManager implements InternalComponentManager {
+  getCapabilities(): InternalComponentCapabilities {
+    return CAPABILITIES;
+  }
+
+  getDebugName({ tagName }: DynamicElementComponentDefinition): string {
+    return tagName;
+  }
+
+  getSelf(): Reference {
+    return NULL_REFERENCE;
+  }
+
+  getDestroyable(): null {
+    return null;
+  }
+}
+
+export const DYNAMIC_ELEMENT_MANAGER = new DynamicElementManager();
+
+class DynamicElementComponentDefinition {
+  constructor(public tagName: string) {}
+}
+
+setInternalComponentManager(DYNAMIC_ELEMENT_MANAGER, DynamicElementComponentDefinition.prototype);
+
+export default function createDynamicElementRef(tagName: Reference) {
+  let lastValue: Maybe<Dict> | string, curriedDefinition: object | string | null;
+
+  return createComputeRef(() => {
+    let value = valueForRef(tagName) as string;
+
+    if (value === lastValue) {
+      return curriedDefinition;
+    }
```

Comment:
I think the code will be easier to write if you do


```ts
let lastValue = null;
let component = null;

return createComputeRef(() => {
  // ...

  if (value === lastValue) {
    return component;
  }

  lastValue = value;
  component = ...;
  
  return component;
});
```

Reply:
I also think that we shouldn't need the "curry" stuff here anymore. If I understand correctly, things are supposed to Just Work™ when you do `<Tag>` where `Tag` is a component. So I don't think we need to go out of our way to "curry nothing" here.

On a related note I think we also try to hide what the return type is since it can leak into user land JS code an ultimately it will be depended on by app code _somehow_ (in this case the `DynamicElementComponentDefinition` class and the fact that it has a `tagName` property on it will accidentally become public api).

So we can do this dance (or there may be a similar pattern elsewhere in glimmer-vm already?):


```ts
// possibly, we can do less work between DEBUG and production, but I think it doesn't really matter
// since this shouldn't be a super common feature?

// not exported, not public
class DynamicElement {
  constructor(readonly tagName: string) {}
}

const DYNAMIC_ELEMENTS = new WeakMap<object, DynamicElement>();

function makeDynamicElement(tagName: string | null): object | null {
  if (tagName === null) {
    return null;
  }

  let component = new DynamicElement(tagName);

  let opaqueValue = Object.freeze({
    // just so you get something nice to look at when logging to the console
    toString() {
      return `(element "${tagName}")`;
    }
  });
  
  setInternalComponentManager(DYNAMIC_ELEMENT_MANAGER, opaqueValue);

  DYNAMIC_ELEMENTS.set(opaqueValue, component);

  return opaqueValue;
}
```